### PR TITLE
Fix the :bug: - once and for all

### LINKED ISSUE #1523: Various GroupedCollectionTableViewSource crashes
There are multiple but similar types of crashes all occuring in `GroupedCollectionTableViewSource`:

**Example**: `NSInternalInconsistencyException: Invalid update: invalid number of rows in section 0. The number of rows contained in an existing section after the update (2) must be equal to the number of rows contained in that section before the update (2), plus or minus the number of rows inserted or deleted from that section (1 inserted, 0 deleted) and plus or minus the number of rows moved into or out of that section (0 moved in, 0 moved out).
`

  [📘 example of stack trace](https://appcenter.ms/orgs/Toggl/apps/Toggl-iOS/crashes/groups/b8d9705249f6cc94759d26d52829e190758c1597)

This happens with variable values. Numbers vary accross different errors (probably depending on the amount of user data).

### LINKED ISSUE #1784: App crashes during animations after pull sync
- [ ] scenario 1
    - create two TEs: X and Y on some date (doesn't matter if you create them on the webapp or in the mobile app)
    - sync the mobile app
    - change the date of both those TEs to a different date in the web app
    - sync the mobile app
    - :boom:
- [ ] scenario 2
    - create two TEs: U and V on 27. 3. (assuming you already don't have no other TEs on that date, otherwise pick some other suitable date, as in the previous case it doesn't matter if it is in the webapp or    the mobile app)
    - sync the mobile app
    - change the date of U to 28. 3. in the web app (also must be emtpy before this step)
    - change the date of V to 29. 3. in the web app (also must be empty before this step)
    - sync the mobile app
    - :boom:
- [ ] scenario 3
    - create two TEs: M and N on two different dates where no other TEs are yet
    - sync the mobile app
    - delete M and N in the web app
    - sync the mobile app
    - :boom:

### PR DESCRIPTION
**Pull request by simonrozsival**
Closes #1523, closes #1784

The main idea behind this PR is that the `GroupedCollectionTableViewSource` handles its own internal state (`ItemsSource`) and makes sure that the displayed sections and rows in the `UITableView` always match the state of the `ItemsSource`. This means that it is not possible to bind anything directly to the `ItemsSource` (as we normally do), but we must rather bind the collection from the VM to the `ObservableCollection` property.

Whenever the `ObservableCollection` changes, the `CollectionChanged` or `OnChildCollectionChanged` events are invoked. We know exactly how the collection was changed (add/delete/replace/...) and we can update both the internal collection and the `UITableView` state - it is important to do this in between `TableView.BeginUpdate()` and `TableView.EndUpdate()` so that we don't get anymore because of the number of rows in the `ItemsSource` and the `UITableView` is different (internal inconsistency).

This PR doesn't try to merge updates into batches, because that would make a lot more changes and it isn't necessary for this fix. We should still do that in the future - it will give use smoother animations. It can certainly wait for 1.5 or 1.6 though. Let's verify this fix and :shipit: ASAP!

Squash whenever

### CODE REVIEW

File: Toggl.Daneel/ViewSources/MainTableViewSource.cs [hunk 1]
```diff
@@ -20,10 +21,16 @@ public IMvxAsyncCommand<TimeEntryViewModel> ContinueTimeEntryCommand
...
+        {
+            get => timeEntriesLogViewSource.ObservableCollection;
+            set => timeEntriesLogViewSource.ObservableCollection = value;
+        }
+
         public override IEnumerable ItemsSource
         {
             get => timeEntriesLogViewSource.ItemsSource;
-            set => timeEntriesLogViewSource.ItemsSource = value;
+            set { throw new InvalidOperationException($"You must bind to the {nameof(ObservableCollection)} and not the {nameof(ItemsSource)}"); }
```
Comment (willsb) [inline Toggl.Daneel/ViewSources/MainTableViewSource.cs]: This makes me happy

File: Toggl.Daneel/ViewSources/GroupedCollectionTableViewSource.cs [hunk 2]
```diff
@@ -163,93 +177,125 @@ private void animateSectionChangesIfPossible(NotifyCollectionChangedEventArgs ar
...
                         break;
                 }
+
+                TableView.EndUpdates();
+            }
+        }
+
+        private void cloneCollection()
+        {
+            internalCollection = new List<TCollection>();
```
Comment (amulware) [inline Toggl.Daneel/ViewSources/GroupedCollectionTableViewSource.cs]: Consider initialising with capacity.

File: Toggl.Daneel/ViewSources/MainTableViewSource.cs (hunk 1 above)
Comment (amulware) [reply Toggl.Daneel/ViewSources/MainTableViewSource.cs]: Why? (curious)

File: Toggl.Daneel/ViewSources/GroupedCollectionTableViewSource.cs [hunk 3]
```diff
@@ -22,18 +22,49 @@ public abstract class GroupedCollectionTableViewSource<TCollection, TItem> : Mvx
...
 
-        protected IEnumerable<MvxObservableCollection<TItem>> GroupedItems
-            => ItemsSource as IEnumerable<MvxObservableCollection<TItem>>;
+        private NestableObservableCollection<TCollection, TItem> observableCollection;
 
-        protected NestableObservableCollection<TCollection, TItem> AnimatableCollection
-            => ItemsSource as NestableObservableCollection<TCollection, TItem>;
+        private readonly IList<TCollection> internalCollection;
+
+        protected IList<TCollection> GroupedItems => internalCollection;
```
Comment (amulware) [inline Toggl.Daneel/ViewSources/GroupedCollectionTableViewSource.cs]: Is this actually exposed to anybody else? Or used at all?

Comment (simonrozsival) [reply Toggl.Daneel/ViewSources/GroupedCollectionTableViewSource.cs]: Yes, it is used 3 times in the 3 derived classes.

Comment (amulware) [reply Toggl.Daneel/ViewSources/GroupedCollectionTableViewSource.cs]: If it's only used for reading, I'd prefer exposing only a cached `internalCollection.AsReadonly()` (could be assigned in the constructor), or is there a reason not to?

File: Toggl.Daneel/ViewSources/GroupedCollectionTableViewSource.cs [hunk 4]
```diff
@@ -22,18 +22,49 @@ public abstract class GroupedCollectionTableViewSource<TCollection, TItem> : Mvx
...
+            }
+        }
 
         protected GroupedCollectionTableViewSource(UITableView tableView, string cellIdentifier, string headerCellIdentifier)
             : base(tableView)
         {
             this.cellIdentifier = cellIdentifier;
             this.headerCellIdentifier = headerCellIdentifier;
 
+            internalCollection = new List<TCollection>();
```
Comment (amulware) [inline Toggl.Daneel/ViewSources/GroupedCollectionTableViewSource.cs]: Initialise field directly?

**Comment by simonrozsival** [comment]
Inserts/deletions aren't `O(n^2)` - they are `O(m * (n + m))`, where `n` represents the number of items in the list and `m` the number of inserted/deleted rows. For small `m`s this converges to `n` and for large `m`s this converges to `m^2`.

We could theoretically achieve `O(n + m)` with a one-pass "merging" of the old data and the changes, but I am not sure if that would be a good idea.

We might use some different data structure - e.g. a heap to store the ordered items of the collection, then the time complexity would be `O(m * ln (n + m))`.

This might be a nice optimization in some of the next releases (1.6?). I think this might be good enough for now.

When I think about it, a much easier way of dealing with the performance might be to set a "limit"  of changes for which we will still animate it (e.g., 20) and for more than that we simply reload the table without any animation.

File: Toggl.Daneel/ViewSources/GroupedCollectionTableViewSource.cs [hunk 5]
```diff
@@ -22,11 +22,40 @@ public abstract class GroupedCollectionTableViewSource<TCollection, TItem> : Mvx
         private readonly object animationLock = new object();
         private readonly List<IDisposable> disposables = new List<IDisposable>();
 
-        protected IEnumerable<MvxObservableCollection<TItem>> GroupedItems
-            => ItemsSource as IEnumerable<MvxObservableCollection<TItem>>;
+        private NestableObservableCollection<TCollection, TItem> observableCollection;
 
-        protected NestableObservableCollection<TCollection, TItem> AnimatableCollection
-            => ItemsSource as NestableObservableCollection<TCollection, TItem>;
+        private List<TCollection> displayedGroupedItems = new List<TCollection>();
```
Comment (amulware) [inline Toggl.Daneel/ViewSources/GroupedCollectionTableViewSource.cs]: This can still be `readonly` right?

**Comment by amulware** [comment]
While we are geeking out about big-O, if we sort a list of updates, and then merge them into the list, it should be `O(n + m * log m)`, right? Anyways. For the future.

**Comment by simonrozsival** [comment]
You are right, I assumed the list of changes was sorted - which it probably wouldn’t be.

File: Toggl.Daneel/ViewSources/MainTableViewSource.cs (hunk 1 above)
Comment (willsb) [reply Toggl.Daneel/ViewSources/MainTableViewSource.cs]: Because that was a very big concern of mine: people binding to the wrong property

Comment (simonrozsival) [reply Toggl.Daneel/ViewSources/MainTableViewSource.cs]: This is still unfortunately only a silent error - MvvmCross catches it and puts it into the application output. But I hope that it is still better than nothing.

**Pull request merged**